### PR TITLE
Partially fix #477

### LINKED ISSUE #477: Assert does not work with ArraySegment
Reported by @xplicit in nunit/nunitv2#39, moved here because we will not fix it in 2.6

This code throws an exception on .NET, while it should be passed. With mono 3.12 it works. The same issue with NUnit v 3.0

     Assert.AreEqual(new ArraySegment<byte>(),new ArraySegment<byte>());

### PR DESCRIPTION
**Pull request by CharliePoole**
ArraySegment still doesn't compare correctly under Silverlight, CF or portable builds because IEnumerable isn't implemented. I decided to leave it and used conditional compilation to keep it out of those builds.

### CODE REVIEW

**Comment by CharliePoole** [comment]
Well, it fails on Appveyor under net-2.0 and net-3.5. I guess I have to check dynamically as I originally said I would!

**Comment by CharliePoole** [comment]
That did it - ready to merge.

**Comment by rprouse** [comment]
Feels a bit icky adding all these special conditions, but the code looks good. Merging.

**Comment by CharliePoole** [comment]
Yes, I think a bunch of refactoring is called for eventually. The key would be getting things as the actual type they are. Have you used 'dynamic'? Would that work in the 4+ runtimes?

**Comment by rprouse** [comment]
dynamic is only supported in .NET 4+ and I am not sure if the DLR is
supported on all other platforms. I doubt it. How do you think dynamic will
help?

On 10 February 2015 at 11:30, CharliePoole <notifications@github.com> wrote:

> Yes, I think a bunch of refactoring is called for eventually. The key
> would be getting things as the actual type they are. Have you used
> 'dynamic'? Would that work in the 4+ runtimes?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nunit/nunit/pull/501#issuecomment-73731359>.
>



-- 

Rob Prouse



I welcome VSRE emails. Learn more at http://vsre.info/

**Comment by CharliePoole** [comment]
Yes, I meant for 4.0 and 4.5 only. Dynamic types allow you to use any methods you like, without the compiler checking that they exist. If they don't, you get a runtime error. This could be helpful where we have an object and actually know what type it is because of some other logic.

I'll do a little experiment with it and post what I get.

File: src/NUnitFramework/framework/Constraints/NUnitEqualityComparer.cs [hunk 1]
```diff
@@ -134,6 +134,9 @@ public bool AreEqual(object x, object y, ref Tolerance tolerance)
             Type xType = x.GetType();
             Type yType = y.GetType();
 
+            Type xGenericTypeDefinition = xType.IsGenericType ? xType.GetGenericTypeDefinition() : null;
```
Comment (oznetmaster) [inline src/NUnitFramework/framework/Constraints/NUnitEqualityComparer.cs]: GetGenericTypeDefinition does not exist in CF

File: src/NUnitFramework/framework/Constraints/NUnitEqualityComparer.cs [hunk 2]
```diff
@@ -134,6 +134,9 @@ public bool AreEqual(object x, object y, ref Tolerance tolerance)
...
+            Type yGenericTypeDefinition = yType.IsGenericType ? yType.GetGenericTypeDefinition() : null;
```
Comment (oznetmaster) [inline src/NUnitFramework/framework/Constraints/NUnitEqualityComparer.cs]: GetGenericTypeDefinition does not exist in CF

**Comment by CharliePoole** [comment]
The code is building in CF 3.5, so GetGenericTypeDefinition appears to exist. I know you can't do much with an open generic type in CF, but this code doesn't do anything with it except to check what the type is.

**Comment by oznetmaster** [comment]
It always throws a Not Supported exception.  Look at the documentation:

 

Type..::.GetGenericTypeDefinition Method

.NET Framework

Supported in: 3.5, 3.0, 2.0

**Comment by oznetmaster** [comment]
I am working on replacement code which will work correctly in CF for this case (checking the type).  In other updates where new code used GetGenericTypeDefinition, it was correctly conditioned out for CF.

**Comment by oznetmaster** [comment]
From the .NET CF source code:

 

public virtual Type GetGenericTypeDefinition()
{
    throw new NotSupportedException();
}

**Comment by CharliePoole** [comment]
And yet all the tests are passing. It looks as if this code must be
executed most of the time when we do an equality comparison, unless
IsGenericType is always returning false as well. I'll have to step through
it.

**Comment by CharliePoole** [comment]
I stepped through and watched it work correctly. I'm using the Windows
Mobile 5.0 Pocket PC R2 emulator.

**Comment by oznetmaster** [comment]
I need to look closer, since I have conditioned out all calls to this method in my CF build, so I never execute the code.  This was based on the documentation stating it was not supported….

**Comment by CharliePoole** [comment]
Would you create an issue for this and assign it to yourself? Something like "research generic support in various CF versions" I guess.

**Comment by oznetmaster** [comment]
Interesting, since the documentation: https://msdn.microsoft.com/en-us/library/System.Type_methods(v=vs.90).aspx clearly indicates that it is not supported on the compact framework (no icon next to the method, and the documentation does not list it as supported).

**Comment by oznetmaster** [comment]
Ok.  Because if this really works all the time, other code that has been eliminated from the CF build (by myself and others) may also actually work.

**Pull request merged**